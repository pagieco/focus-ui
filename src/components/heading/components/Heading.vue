<script>

export const ElementTypes = {
  H1: 'h1',
  H2: 'h2',
  H3: 'h3',
  H4: 'h4',
  H5: 'h5',
  H6: 'h6',
  P: 'p',
};

export default {
  name: 'focus-heading',

  props: {
    element: {
      type: String,
      default: ElementTypes.H2,
      validator: val => Object
        .values(ElementTypes)
        .includes(val),
    },
  },
};

</script>

<template>
  <h2 :is="element" class="heading">
    <slot />
  </h2>
</template>
