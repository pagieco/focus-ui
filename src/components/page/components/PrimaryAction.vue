<script>

import { pick } from 'lodash';
import { Button } from '../../button';

export default {
  components: { Button },

  props: {
    ...Button.props,

    content: {
      type: String,
    },
  },

  computed: {
    buttonProps() {
      return pick(this.$props, Object.keys(Button.props));
    },
  },
};

</script>

<template>
  <Button v-bind="buttonProps">
    {{ content }}
  </Button>
</template>
