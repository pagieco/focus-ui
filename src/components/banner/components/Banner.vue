<script>

import Heading from '../../heading/components/Heading.vue';
import { filterClassList, uuidv4 } from '../../../utilities';
import PrimaryAction from './PrimaryAction.vue';
import { Icon } from '../../icon';

export const Status = {
  Success: 'success',
  Info: 'info',
  Warning: 'warning',
  Critical: 'critical',
};

export default {
  props: {
    /**
     * Whether or not the banner can be dismissed.
     */
    dismissable: {
      type: Boolean,
      default: true,
    },

    /**
     * The primary action for the banner.
     */
    primaryAction: {
      type: Object,
    },

    /**
     * Sets the status of the banner.
     */
    status: {
      type: String,
      default: null,
      validator: val => Object
        .values(Status)
        .concat([null])
        .includes(val),
    },

    /**
     * The title content for the banner.
     */
    title: {
      type: String,
    },
  },

  data() {
    return {
      componentId: null,
    };
  },

  components: {
    Icon,
    Heading,
    PrimaryAction,
  },

  mounted() {
    this.componentId = uuidv4();
  },

  computed: {
    classList() {
      return filterClassList([
        'banner',
        [`banner--status-${this.status}`],
      ]);
    },

    bannerRole() {
      return this.status === Status.Warning || this.status === Status.Critical
        ? 'alert'
        : 'status';
    },

    ariaDescribedBy() {
      return this.title
        ? `BannerTitle_${this.componentId}`
        : `BannerContent_${this.componentId}`;
    },
  },
};

</script>

<template>
  <div tabindex="0" :class="classList" :role="bannerRole" :aria-describedby="ariaDescribedBy">
    <div class="banner__ribbon"></div>

    <div class="banner__dismiss" v-if="dismissable">
      <button class="banner__dismiss__button" @click="$emit('dismiss')">
        <Icon name="times" />
      </button>
    </div>

    <div>
      <div class="banner__heading" v-if="title">
        <Heading :id="`BannerTitle_${componentId}`">{{ title }}</Heading>
      </div>
      <div class="banner__content" :id="`BannerContent_${componentId}`">
        <slot />

        <div class="banner__actions" v-if="primaryAction">
          <PrimaryAction v-bind="{ ...primaryAction }" />
        </div>
      </div>
    </div>
  </div>
</template>
