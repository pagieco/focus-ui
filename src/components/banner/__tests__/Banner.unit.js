import { shallowMount } from '@vue/test-utils';
import Banner, { Status } from '../components/Banner.vue';
import Heading from '../../heading/components/Heading.vue';
import PrimaryAction from '../components/PrimaryAction.vue';

const shallowWrap = propsData => shallowMount(Banner, { propsData });

describe('Banner', () => {
  it('is a valid vue component', () => {
    expect(Banner).toBeAComponent();
  });

  it('matches the snapshot', () => {
    expect(shallowMount(Banner).element).toMatchSnapshot();
  });

  it('uses the primary-action prop', () => {
    const primaryAction = { content: 'Gaaf' };
    const wrapper = shallowWrap({ primaryAction });

    expect(wrapper.contains(PrimaryAction)).toBeTruthy();
  });

  it('uses the status prop', () => {
    const wrapper = shallowWrap({ status: Status.Info });

    expect(wrapper.classes('banner--status-info')).toBeTruthy();
  });

  it('validates the status prop', () => {
    const prop = Banner.props.status;

    expect(prop.validator).toBeInstanceOf(Function);
    expect(prop.validator('success')).toBeTruthy();
    expect(prop.validator('info')).toBeTruthy();
    expect(prop.validator('warning')).toBeTruthy();
    expect(prop.validator('critical')).toBeTruthy();
    expect(prop.validator('dangerous')).toBeFalsy();
  });

  it('uses the title prop', () => {
    const title = 'Banner Title';
    const wrapper = shallowWrap({ title });

    expect(wrapper.find(Heading).text()).toBe(title);
  });

  it('generates the correct role attribute', () => {
    expect(shallowWrap({ status: Status.Success }).attributes('role')).toBe('status');
    expect(shallowWrap({ status: Status.Info }).attributes('role')).toBe('status');
    expect(shallowWrap({ status: Status.Warning }).attributes('role')).toBe('alert');
    expect(shallowWrap({ status: Status.Critical }).attributes('role')).toBe('alert');
  });

  it('triggers the dismiss event', () => {
    const wrapper = shallowWrap();

    wrapper.find('.banner__dismiss__button').trigger('click');

    expect(wrapper.emitted('dismiss')).not.toBeUndefined();
  });
});
