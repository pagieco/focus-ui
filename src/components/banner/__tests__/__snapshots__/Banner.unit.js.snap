// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner matches the snapshot 1`] = `
<div
  aria-describedby="BannerContent_null"
  class="banner banner--status-null"
  role="status"
  tabindex="0"
>
  <div
    class="banner__ribbon"
  />
   
  <div
    class="banner__dismiss"
  >
    <button
      class="banner__dismiss__button"
    >
      <icon-stub
        name="times"
      />
    </button>
  </div>
   
  <div>
    <!---->
     
    <div
      class="banner__content"
      id="BannerContent_null"
    >
       
      <!---->
    </div>
  </div>
</div>
`;
